testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakContainerEventsController.java

```diff
@@ -66,8 +79,98 @@ public void execute(@Observes(precedence = 3) AfterClass event) {
         } finally {
             container.fire(new StopClassContainers());
         }
+        if (event.getTestClass().isAnnotationPresent(RestartContainer.class)) {
+            afterContainerStop(event.getTestClass().getAnnotation(RestartContainer.class).value());
+        }
     }
 
+    @Override
+    public void execute(BeforeClass event) {
+        if (event.getTestClass().isAnnotationPresent(RestartContainer.class)) {
+            container.fire(new StopManualContainers());
+            container.fire(new StopSuiteContainers());
+
+            beforeContainerStart(event.getTestClass().getAnnotation(RestartContainer.class).value());
+
+            container.fire(new StartClassContainers());
+            container.fire(new StartSuiteContainers());
+        }
+        super.execute(event);
+    }
+
+    /**
+     * Actions to take after the original container stops.
+     * @param value passed from annotation {@link RestartContainer} which triggered this behaviour
+     */
+    protected void afterContainerStop(String value) {
+        copyKeycloakAddUserFile();
+    }
+
+    /**
+     * Actions to take before new container starts.
+     * @param value passed from annotation {@link RestartContainer} which triggered this behaviour
+     */
+    protected void beforeContainerStart(String value) {
+        removeKeycloakAddUserFile();
```

Comment:
I wonder if to add the condition like "if (WELCOME_PAGE_TEST.equals(value)) {" (same for afterContainerStopMethod) and use the annotation on WelcomePageTest like @RestartContainer(value = WELCOME_PAGE_TEST) ? If I understand correctly, the value can be left empty now as WelcomePAgeTest is the only test, which uses it right now, however if more tests are added in the future, it would need to be changed anyway. I am not 100% sure about this, so leaving to you how to do it in the end :)

Reply:
Addressed: used value and set default behavior.